File: Priority_Queue/1705. Maximum Number of Eaten Apples/1705. Maximum Number of Eaten Apples.cpp
using PII = pair<int, int>; // {date, num}
class Solution {
public:
    int eatenApples(vector<int>& apples, vector<int>& days) {
        int n = apples.size();
        int rets = 0;
        priority_queue<PII, vector<PII>, greater<PII>>pq;
        
        int day = 0;
        while (day < n || !pq.empty()) {
            // eat the earliest rot
            while (!pq.empty() && pq.top().first <= day) { // rotten apple
                pq.pop();
            }
            
            if (day < n && apples[day] != 0) {
                pq.push({day+days[day], apples[day]});
            }
            
            if (!pq.empty()) {
                auto cur = pq.top();
                pq.pop();
                rets++;
                if (cur.second >= 2) {
                    pq.push({cur.first, cur.second-1});
                }
            }
            day++;
        }
        
        return rets;
    }
};

File: Priority_Queue/1705. Maximum Number of Eaten Apples/1705. Maximum Number of Eaten Apples.java
class Solution {
    // pq
    public int eatenApples(int[] apples, int[] days) {
        int n = apples.length;
        int rets = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>( (a, b) -> (a[0]-b[0]) );
        
        int day = 0;
        while (day < n || !pq.isEmpty()) {
            // eat the earliest rot
            while (!pq.isEmpty() && pq.peek()[0] <= day) { // rotten apple
                pq.poll();
            }
            
            if (day < n && apples[day] != 0) {
                pq.offer(new int[] {day+days[day], apples[day]});
            }
            
            if (!pq.isEmpty()) {
                int[] cur = pq.poll();
                rets++;
                if (cur[1] >= 2) {
                    pq.offer(new int[] {cur[0], cur[1]-1});
                }
            }
            day++;
        }
        
        return rets;
    }
}

// NlgN
// heap
// n = 4e4
